Encoding dictionary

quinmars opened the pull request:
This is the second iteration of #131. I have toyed with the idea a little further, what I do not like is that the `EncodingMap` has two phases in it's lifetime. In the first it gets filled or composed and in the second phase it works as an dictionary. Those two phases do not fit very well into one data type or interface. I propose here a different way: splitting those two phases into objects.

### First phase: encoding table

Outside of the channel the user can compose the types and node id pairs with encoding tables. This is not a type/interface of its own but just another name for `IEnumerable<(ExpandedNodeId,Type)>`. There are some helper types to create such tables: `BinaryEncodingTable`, `CustomEncodingTable`, and `MergedEncodingTable`. So a user can compose the encoding table with:

```csharp
IEnumerable<(ExpandedNodeId,Type)> encodingTable = new MergedEncodingTable
{
    new BinaryEncodingTable(),
    new CustomEncodingTable
    {
        (new ExpandedNodeId(...), typeof(MyType))
    }
}
```

Or if he prefers to use an attribute based mapping:


```csharp
IEnumerable<(ExpandedNodeId,Type)> encodingTable = new MergedEncodingTable
{
    new BinaryEncodingTable(),
    new CustomEncodingTable<BinaryEncodingIdAttribute>
    {
        typeof(MyType1),
        typeof(MyType2),
        ...
    }
}
```

The user can even define his own attribute, for example `BREncodingIdAttribute` and `OmronEncodingIdAttribute`, this would solve #129. And if you do not want to list all type, you can also let the runtime scan the whole assembly for you:

```csharp
IEnumerable<(ExpandedNodeId,Type)> encodingTable = new MergedEncodingTable
{
    new BinaryEncodingTable(),
    new CustomEncodingTable<BinaryEncodingIdAttribute>(assembly)
}
```

I thought about ommitting the binary encoding table and making it default. There are two reasons why I decided against it. First, it allows the user two define it's own types. Of course there are many types `UaClient` relies on, nontheless the user could define derivates and it should just work. Second, the user can/must define an alterative encoding table for example for `Json`, once `UaClient` supports this. The latter is rather a counterargument. I actually do not understand why the OPC UA foundation specifies different node IDs for the same type, just because of the encoding. That mixes two concepts that should be orthogonal, but that is another story. Anyway, if you think we should stick to a default handling there I can adjust the PR.

### Second phase: encoding dictionary

Once the user has composed the encoding table, he pass it to the channel constructor. There, a dictionary is build based on the table. Unlike in #131 the dictionary is not injected by the user nor is it mutable. So infact, we do not need there an interface, we can simply use a class, that is `EncodingDictionary`. When the channel receives the namespace URIs, a new dictionary is build and used upon.


This PR does not introduce the `IEncodingContext` interface, yet. There are also unit tests, null checking and documentation missing. I first wanted your reaction before I start to add those things.

There are also some open points:

 1. Do you like the name _encoding table_? I think it fits, but I'm not absolutely sure. Alternatives could be _type set_ or _node set_.
 2. I will write some comment as a code review to highlight some other points.

quinmars commented on UaClient/ServiceModel/Ua/CustomEncodingTable.cs:
```diff
@@ -0,0 +1,91 @@
+                        let info = type.GetTypeInfo()
+                        where info.ImplementedInterfaces.Contains(typeof(IEncodable))
+                        let attr = info.GetCustomAttribute<TAttribute>(false)
+                        where attr != null
```
Should we fail here? Currently, I skip those types silently.

awcullen replied:
would be better to fail here

quinmars replied:
It throws now.

quinmars commented on UaClient/ServiceModel/Ua/CustomEncodingTable.cs:
```diff
@@ -0,0 +1,91 @@
+                        let info = type.GetTypeInfo()
+                        where info.ImplementedInterfaces.Contains(typeof(IEncodable))
+                        let attr = info.GetCustomAttribute<TAttribute>(false)
+                        where attr != null
```
Here it is correct, to skip types, because not all types of an assembly must be annotated.

awcullen replied:
I agree we should skip here.

quinmars commented on UaClient/ServiceModel/Ua/CustomEncodingTable.cs:
```diff
@@ -0,0 +1,91 @@
+            if (info.ImplementedInterfaces.Contains(typeof(IEncodable)))
+            {
+                var attr = info.GetCustomAttribute<TAttribute>(false);
+                if (attr != null)
```
But maybe here we should fail?

awcullen replied:
better to fail here

quinmars replied:
It throws now.

quinmars commented on UaClient/ServiceModel/Ua/EncodingDictionary.cs:
```diff
@@ -0,0 +1,76 @@
+                    var i = namespaceUris.IndexOf(nsu!);
+                    if (i == -1)
+                    {
+                        continue;
```
I intentionally do not use `ExpandableNodeId.ToNodeId` because it could silently convert a node with unknown namespace (not in the URI list) to a `ns=0` node, and in worst case overrride a default type.

quinmars commented on UaClient/ServiceModel/Ua/EncodingDictionary.cs:
```diff
@@ -0,0 +1,76 @@
+                var nsu = nodeId.NamespaceUri;
+                if (!string.IsNullOrEmpty(nsu))
+                {
+                    continue;
```
This constructor is called before we retrieved the URI list. Should we perform more checks to be sure that this node id/type pair does (not) belongs to the OPC UA namespace?

awcullen commented:
When adding types and encodingIds, we should throw if the type or encodingId is already present in the dictionary.  I want to be sure we don't allow two ids for one type.

quinmars commented:
> When adding types and encodingIds, we should throw if the type or encodingId is already present in the dictionary. I want to be sure we don't allow two ids for one type.

I'm using now `Dictionary.Add` instead of the indexer.

quinmars commented:
I added now the outstanding null checks, unit tests and the documentation. Looking now back, the total changes to the public API are less than planned, but I like the outcome even more than the original plan. The secure channel is now less dependend on the binary encoding, we moved the dictionary creation logic out of the channel, so it is now testable and offer the user easy-to-use methods to build encoding table.

awcullen commented:
I invite you to comment on PR #158. In #158, does the singleton TypeLibrary cause any issues?  Is it acceptable that type dictionaries are read-only and can not be added to after initialization?

quinmars commented:
@awcullen Thank you for the invitation. I already looked through that PR, but I haven't found the time yet to write my ideas down. I'll do that tomorrow.
